**Mount Google Drive to look up for csv file**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**1. Analyze variables, fill in missing values, normalize extreme values.**

In [ ]:
import pandas as pd 

#CARGAR EL ARCHIVO CON LOS DATOS AJUSTADOS EN EXCEL
df=pd.read_csv(r'/content/drive/MyDrive/Datos analizados para modelo data science.csv')
df.head()


In [ ]:
df['name'] = df['name'].str.replace(r"_"," ")
df['name'] = df['name'].str.replace(r"\W"," ")
df['model'] = df['model'].str.replace(r"\W"," ")
df['name'] = df['name'].str.replace(r"  "," ")
df['name'] = df['name'].str.replace(r"  "," ")
df['name'] = df['name'].str.upper()
df['seller'] = df['seller'].str.title()
df['offerType'] = df['offerType'].str.title()
df['abtest'] = df['abtest'].str.title()
df['vehicleType'] = df['vehicleType'].str.title()
df['gearbox'] = df['gearbox'].str.title()
df['model'] = df['model'].str.title()
df['fuelType'] = df['fuelType'].str.title()
df['brand'] = df['brand'].str.title()
df['notRepairedDamage'] = df['notRepairedDamage'].str.title()
df['dateCrawled'] = pd.to_datetime(df['dateCrawled'],yearfirst=True, errors='coerce')
df['dateCreated'] = pd.to_datetime(df['dateCreated'],yearfirst=True, errors='coerce')
df['lastSeen'] = pd.to_datetime(df['lastSeen'],yearfirst=True, errors='coerce')
df['notRepairedDamage'] = df['notRepairedDamage'].fillna('No')  
df['name'] = df['name'].str.strip()

df = df.fillna('Not Available')

df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until


dateCrawled                                               name  \
0 2016-03-26 14:57:00  BMW 320D DPFXENON TEMPOMAT SITZHEIZUNG PDC KLI...   
1 2016-03-23 20:57:00                      RENAULT CLIO MIT NAGELNEUEN T   
2 2016-03-19 18:56:00                                VW LUPO 1 1L T 3 18   
3 2016-03-30 12:51:00             OPEL ASTRA CARAVAN SPORT SEHR GEPFLEGT   
4 2016-03-26 21:36:00              FORD FOCUS TURNIER 1 6 TDCI DPF TREND   

   seller offerType  price   abtest vehicleType  yearOfRegistration gearbox  \
0  Privat   Angebot  10499  Control   Limousine                2006  Manual   
1  Privat   Angebot   1199     Test  Kleinwagen                2001  Manual   
2  Privat   Angebot   2750     Test  Kleinwagen                1999  Manual   
3  Privat   Angebot   3500  Control       Kombi                2004  Manual   
4  Privat   Angebot   7500  Control       Kombi                2012  Manual   

   powerPS (Horse Power)  ...       brand  notRepairedDamage  dateCreated  \
0                    163  ...         Bmw                 No   2016-03-26   
1                      0  ...     Renault                 No   2016-03-23   
2                     50  ...  Volkswagen                 No   2016-03-19   
3                    125  ...        Opel                 No   2016-03-30   
4                    116  ...        Ford                 No   2016-03-26   

  nrOfPictures postalCode            lastSeen    Unnamed: 20    Unnamed: 21  \
0            0      76661 2016-04-06 02:46:00  Not Available  Not Available   
1            0      10967 2016-04-06 02:46:00  Not Available  Not Available   
2            0      53757 2016-03-23 08:15:00  Not Available  Not Available   
3            0       4329 2016-04-01 12:50:00  Not Available  Not Available   
4            0      91189 2016-04-02 14:16:00  Not Available  Not Available   

     Unnamed: 22    Unnamed: 23  
0  Not Available  Not Available  
1  Not Available  Not Available  
2  Not Available  Not Available  
3  Not Available  Not Available  
4  Not Available  Not Available  

[5 rows x 24 columns]

In [ ]:
print(len(df.name.unique()))
print(df.name.unique())

92984
['BMW 320D DPFXENON TEMPOMAT SITZHEIZUNG PDC KLIMA MFL'
 'RENAULT CLIO MIT NAGELNEUEN T' 'VW LUPO 1 1L T 3 18' ...
 'BMW 330D TOURING SPECIAL EDITION' 'ESCORT 16 V MIT 90 PS FAEHRT KLASSE'
 'VOLKSWAGEN SHARAN BLUEMOTION TRENDLINE']


In [ ]:
print(df.seller.unique())


['Privat']


In [ ]:
print(df.offerType.unique())

['Angebot' 'Gesuch']


In [ ]:
print(min(df.price.unique()))
print(max(df.price.unique()))
print(df.price.unique())

0
99999999
[10499  1199  2750 ... 27333  2598 21959]


In [ ]:
print(df.abtest.unique())

['Control' 'Test']


In [ ]:
print(df.vehicleType.unique())

['Limousine' 'Kleinwagen' 'Kombi' 'Suv' 'Bus' 'Not Available' 'Cabrio'
 'Coupe' 'Andere']


In [ ]:
print(df.yearOfRegistration.unique())

[2006 2001 1999 2004 2012 2002 1998 2018 2005 2011 2000 2007 2003 2008
 2013 2016 2009 1994 2010 1967 1993 2015 1997 1992 1980 2014 2017 1995
 1988 1991 1990 1996 1984 1989 1987 1978 1985 1966 1981 1965 1970 1968
 1974 1986 1982 1976 1931 1973 1972 1969 1000 1983 1979 1971 1962 1910
 1975 1961 7500 1977 1800 1960 1929 1957 1952 1964 1955 1956 1919 1963
 5555 9999 9000 7777 1958 1953 1950 5600 1959 2019 1600 1111 1954 1932
 1947 1939 2222 1943 1039 1951 3000 5000 1938 6000 1942 9996 1928 1933
 1934 1948 1937 1300 2800 1930 8455 7000 1915 4500 1500 1935 1920 1602
 8888 7800 9229 1927 1936 1946 7100 2500 8200 1923]


In [ ]:
print(df.gearbox.unique())

['Manual' 'Automatik' 'Not Available']


In [ ]:
print(df['powerPS (Horse Power)'].unique())

[  163     0    50   125   116   245   105   115   135   170   182    86
   140   232    64   101   136   126    95   177   165    60   109   150
   122    75    90   131   132   143    69   128   102   100   156    54
   326    85   190    73    68   118    74   120   112    45   110   103
   231    72    92   306    97   204   160    65   230    80   218   200
   113   210   145    82   215   349   224   275    52   239   280   174
   241   173    88    55    58   235   167   193   129   250   313    33
   192   130   225   107    53    77   185   211    41    56    98   152
   184   220   186   114    59   265    71   104   320   901   450   355
   141   226  1800   179    84    26   134   286   330    70   133   197
   420    76   155   213   299   334   228   272   223   385    67   111
   147   500   205   273   300    40   333    91   252   175    61    44
   305    38    94   180    48   340   270   400    63    78    79   117
   139   457   209    99   256   144  1317   354   

In [ ]:
print(len(df.model.unique()))
df.model.unique()

251


array(['3Er', 'Clio', 'Lupo', 'Astra', 'Focus', 'Q5', '147', 'Espace',
       'Andere', 'Z_Reihe', 'C_Max', 'Yaris', 'A3', 'A6', 'Corsa', 'Polo',
       'Passat', '2_Reihe', 'Auris', 'Golf', 'A5', 'Ceed', '1Er', nan,
       'Accord', '3_Reihe', 'C_Klasse', 'Caddy', 'Micra', 'Cayenne',
       'Meriva', 'Zafira', 'Vectra', 'A_Klasse', 'A4', 'Fox', 'Galaxy',
       'One', 'Transporter', '5Er', 'Laguna', 'Touran', 'Fiesta',
       'Qashqai', 'Tiguan', 'Berlingo', 'Stilo', 'Mondeo', 'Scenic',
       'Corolla', 'Yeti', 'Cr_Reihe', 'Q7', 'E_Klasse', 'Aygo', 'Eos',
       'A1', 'Fortwo', 'Jimny', 'Beetle', 'Slk', 'Colt', 'Megane',
       'Omega', 'C1', 'Civic', '90', 'Mx_Reihe', 'Ducato', 'M_Klasse',
       'A8', 'Cooper', 'X_Type', 'Fabia', 'Swift', 'Transit', 'Leon',
       'Rio', 'Punto', 'Panda', 'Tucson', 'Tt', 'Alhambra', 'Kangoo',
       'Touareg', 'Twingo', 'Kuga', '500', 'Range_Rover_Evoque', '80',
       'X_Reihe', 'C4', 'Seicento', 'Cuore', 'C3', 'Scirocco', 'Ibiza',
       'Primera

In [ ]:
print(df.kilometer.unique())

[ 90000 150000  40000  50000 100000 125000  70000   5000  80000  30000
  60000  20000  10000]


In [ ]:
#Cero = no registro
print(df['monthOfRegistration'].unique())

[ 1  4 10  8  5  6  0  7  3  9 12  2 11]


In [ ]:
print(df.fuelType.unique())

['Diesel' 'Benzin' 'Lpg' nan 'Cng' 'Hybrid' 'Andere' 'Elektro']


In [ ]:
print(df.brand.unique())

['Bmw' 'Renault' 'Volkswagen' 'Opel' 'Ford' 'Audi' 'Alfa_Romeo' 'Toyota'
 'Peugeot' 'Kia' 'Sonstige_Autos' 'Honda' 'Saab' 'Mercedes_Benz' 'Nissan'
 'Porsche' 'Fiat' 'Mini' 'Citroen' 'Volvo' 'Skoda' 'Mitsubishi' 'Mazda'
 'Smart' 'Suzuki' 'Jaguar' 'Seat' 'Hyundai' 'Jeep' 'Land_Rover' 'Daihatsu'
 'Rover' 'Chevrolet' 'Chrysler' 'Lada' 'Trabant' 'Dacia' 'Lancia' 'Daewoo'
 'Subaru']


In [ ]:
print(df.notRepairedDamage.unique())

['No' 'Yes']


In [ ]:
normalized_price=(df['price']-df['price'].mean())/(df['price'].max() - df['price'].min())
print(normalized_price)

0         6.858556e-07
1        -9.231415e-05
2        -7.680415e-05
3        -6.930415e-05
4        -2.930414e-05
              ...     
140565   -4.580414e-05
140566   -7.930415e-05
140567    1.169586e-05
140568   -9.920415e-05
140569    5.419586e-05
Name: price, Length: 140570, dtype: float64


In [ ]:
normalized_year=(df['yearOfRegistration']-df['yearOfRegistration'].mean())/(df['yearOfRegistration'].max() - df['yearOfRegistration'].min())
print(normalized_year)

0         0.000155
1        -0.000400
2        -0.000622
3        -0.000067
4         0.000822
            ...   
140565   -0.000067
140566   -0.000622
140567    0.000822
140568   -0.000845
140569    0.000711
Name: yearOfRegistration, Length: 140570, dtype: float64


In [ ]:
normalized_km=(df['kilometer']-df['kilometer'].mean())/(df['kilometer'].max() - df['kilometer'].min())
print(normalized_km)

0        -0.245532
1         0.168261
2        -0.590359
3         0.168261
4         0.168261
            ...   
140565    0.168261
140566    0.168261
140567    0.168261
140568   -0.176566
140569   -0.176566
Name: kilometer, Length: 140570, dtype: float64


In [ ]:
#FACTORIZE (Categorica a numerica)
Fact_abtest = pd.factorize(df.abtest)
Fact_abtest

(array([0, 1, 1, ..., 0, 1, 1]), Index(['Control', 'Test'], dtype='object'))

In [ ]:
Fact_vehicletype = pd.factorize(df.vehicleType,)
Fact_vehicletype

(array([0, 1, 1, ..., 0, 0, 4]),
 Index(['Limousine', 'Kleinwagen', 'Kombi', 'Suv', 'Bus', 'Cabrio', 'Coupe',
        'Andere'],
       dtype='object'))

In [ ]:
Fact_gearbox = pd.factorize(df.gearbox)
Fact_gearbox

(array([0, 0, 0, ..., 0, 0, 0]),
 Index(['Manual', 'Automatik'], dtype='object'))

In [ ]:
Fact_model = pd.factorize(df.model)
Fact_model

(array([  0,   1,   2, ...,  89, 137, 111]),
 Index(['3Er', 'Clio', 'Lupo', 'Astra', 'Focus', 'Q5', '147', 'Espace',
        'Andere', 'Z_Reihe',
        ...
        'Crossfire', 'I3', 'Kappa', 'Gl', '200', 'B_Max', 'Serie_1', 'Serie_3',
        'Discovery_Sport', 'Rangerover'],
       dtype='object', length=250))

In [ ]:
Fact_brand = pd.factorize(df.brand)
print(Fact_brand)

(array([0, 1, 2, ..., 0, 4, 2]), Index(['Bmw', 'Renault', 'Volkswagen', 'Opel', 'Ford', 'Audi', 'Alfa_Romeo',
       'Toyota', 'Peugeot', 'Kia', 'Sonstige_Autos', 'Honda', 'Saab',
       'Mercedes_Benz', 'Nissan', 'Porsche', 'Fiat', 'Mini', 'Citroen',
       'Volvo', 'Skoda', 'Mitsubishi', 'Mazda', 'Smart', 'Suzuki', 'Jaguar',
       'Seat', 'Hyundai', 'Jeep', 'Land_Rover', 'Daihatsu', 'Rover',
       'Chevrolet', 'Chrysler', 'Lada', 'Trabant', 'Dacia', 'Lancia', 'Daewoo',
       'Subaru'],
      dtype='object'))


In [ ]:
Fact_Damage = pd.factorize(df.notRepairedDamage)
Fact_Damage

(array([0, 0, 0, ..., 0, 0, 0]), Index(['No', 'Yes'], dtype='object'))

In [ ]:
import numpy as np
df.Delta =  (df.lastSeen - df.dateCreated)/np.timedelta64(1,'D')
df.Delta

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  


0         11.115278
1         14.115278
2          4.343750
3          2.534722
4          7.594444
            ...    
140565    10.011806
140566     1.929167
140567     3.699306
140568     4.572917
140569    18.929167
Length: 140570, dtype: float64

In [ ]:
obj_df = df.select_dtypes(include=['object']).copy()
obj_df.head()

name  seller offerType  \
0  BMW 320D DPFXENON TEMPOMAT SITZHEIZUNG PDC KLI...  Privat   Angebot   
1                      RENAULT CLIO MIT NAGELNEUEN T  Privat   Angebot   
2                                VW LUPO 1 1L T 3 18  Privat   Angebot   
3             OPEL ASTRA CARAVAN SPORT SEHR GEPFLEGT  Privat   Angebot   
4              FORD FOCUS TURNIER 1 6 TDCI DPF TREND  Privat   Angebot   

    abtest vehicleType gearbox  model fuelType       brand notRepairedDamage  \
0  Control   Limousine  Manual    3Er   Diesel         Bmw                No   
1     Test  Kleinwagen  Manual   Clio   Benzin     Renault                No   
2     Test  Kleinwagen  Manual   Lupo   Benzin  Volkswagen                No   
3  Control       Kombi  Manual  Astra   Benzin        Opel                No   
4  Control       Kombi  Manual  Focus   Diesel        Ford                No   

     Unnamed: 20    Unnamed: 21    Unnamed: 22    Unnamed: 23  
0  Not Available  Not Available  Not Available  Not Available  
1  Not Available  Not Available  Not Available  Not Available  
2  Not Available  Not Available  Not Available  Not Available  
3  Not Available  Not Available  Not Available  Not Available  
4  Not Available  Not Available  Not Available  Not Available

In [ ]:
from sklearn import datasets, linear_model
import matplotlib.pyplot as plt

In [ ]:
#del obj_df
obj_df["offerType"] = obj_df["offerType"].astype('category')
obj_df["abtest"] = obj_df["abtest"].astype('category')
obj_df["vehicleType"] = obj_df["vehicleType"].astype('category')
obj_df["gearbox"] = obj_df["gearbox"].astype('category')
obj_df["fuelType"] = obj_df["fuelType"].astype('category')
obj_df["brand"] = obj_df["brand"].astype('category')
obj_df["notRepairedDamage"] = obj_df["notRepairedDamage"].astype('category')
#obj_df["Normalized price"] = df['price']


obj_df["Normalized price"] = (df['price']-df['price'].mean())/(df['price'].max() - df['price'].min())
obj_df["Normalized year"] = (df['yearOfRegistration']-df['yearOfRegistration'].mean())/(df['yearOfRegistration'].max() - df['yearOfRegistration'].min())
obj_df['Normalized km'] = (df['kilometer']-df['kilometer'].mean())/(df['kilometer'].max() - df['kilometer'].min())

obj_df["offerType"] = obj_df["offerType"].cat.codes
obj_df["abtest"] = obj_df["abtest"].cat.codes
obj_df["vehicleType"] = obj_df["vehicleType"].cat.codes
obj_df["gearbox"] = obj_df["gearbox"].cat.codes
obj_df["fuelType"] = obj_df["fuelType"].cat.codes
obj_df["brand"] = obj_df["brand"].cat.codes
obj_df["notRepairedDamage"] = obj_df["notRepairedDamage"].cat.codes

#obj_df= obj_df.drop(columns=['model'])
#obj_df= obj_df.drop(columns=['Unnamed: 20','Unnamed: 21','Unnamed: 22','Unnamed: 23'])
#obj_df = obj_df.reset_index()
obj_df = obj_df[np.all(np.isfinite(x), axis=1)]
obj_df.head()

level_0  index                                               name  seller  \
0        0      0  BMW 320D DPFXENON TEMPOMAT SITZHEIZUNG PDC KLI...  Privat   
1        1      1                      RENAULT CLIO MIT NAGELNEUEN T  Privat   
2        2      2                                VW LUPO 1 1L T 3 18  Privat   
3        3      3             OPEL ASTRA CARAVAN SPORT SEHR GEPFLEGT  Privat   
4        4      4              FORD FOCUS TURNIER 1 6 TDCI DPF TREND  Privat   

   offerType  abtest  vehicleType  gearbox  model  fuelType  brand  \
0          0       0            6        1    3Er         3      2   
1          0       1            4        1   Clio         1     27   
2          0       1            4        1   Lupo         1     38   
3          0       0            5        1  Astra         1     24   
4          0       0            5        1  Focus         3     10   

   notRepairedDamage  Normalized price  Normalized year  Normalized km  
0                  0      6.858556e-07         0.000155      -0.245532  
1                  0     -9.231415e-05        -0.000400       0.168261  
2                  0     -7.680415e-05        -0.000622      -0.590359  
3                  0     -6.930415e-05        -0.000067       0.168261  
4                  0     -2.930414e-05         0.000822       0.168261

In [ ]:
x = obj_df[['brand','Normalized price', 'Normalized year','Normalized km','offerType', 'abtest', 'vehicleType','gearbox','fuelType','notRepairedDamage']].values
y = df.Delta

In [ ]:
regr = linear_model.LinearRegression()
regr.fit(x, y)
print(regr.coef_)

[-0.00971694  1.5109015  -0.97643034 -2.43542799 -3.4048081   0.14143396
 -0.0714161  -0.83340152 -0.05219783 -1.09422732]


MULTIPLE LINEAR REGRESSION MODEL TO PREDICT THE NUMBER OF DAYS A CAR WILL STAY ON THE SITE UNTIL IS SOLD IS GIVEN BY 

Y (number of days) = -0.00971694 * BRAND + 1.5109015 * NORMALIZED PRICE - 0.97643034 * NORMALIZED YEAR - 2.43542799 * NORMALIZED KM - 3.4048081 * OFFER TYPE + 0.14143396 * ABTEST - 0.0714161 * VEHICLE TYPE - 0.83340152 * GEARBOX - 0.05219783 * FUEL TYPE - 1.09422732 * NOT REPAIRED DAMAGE
